posts/2018-05-09-tagless-final-streaming.md

```diff
@@ -0,0 +1,285 @@
+
+Well, it seems that our options are either adding another higher-kinded parameter `G[_]` to our algebra or just define an abstract member `G[_]`. So let's go with the first one:
+
+```scala
+trait ItemRepository[F[_], G[_]] {
+  def findAll: G[Item]
+  def find(name: ItemName): F[Option[Item]]
+  def save(item: Item): F[Unit]
+  def remove(name: ItemName): F[Unit]
+}
+```
+
+Great! This looks good so far.
+
+### Streaming support interpreter
+
+Now let's write a new `PostgreSQL` interpreter with streaming support:
+
+```scala
+import doobie.implicits._
+import doobie.util.transactor.Transactor
+import fs2.Stream
+
+// Doobie implementation (not fully implemented, what matters here are the types).
+class StreamingItemRepository[F[_]](xa: Transactor[F])
+                                   (implicit F: Sync[F]) extends ItemRepository[F, Stream[F, ?]] {
+
+  override def findAll: Stream[F, Item] = sql"select name, price from items"
+                                           .query[Item]
+                                           .stream
+                                           .transact(xa)
+
+  override def find(name: ItemName): F[Option[Item]] = F.pure(None)
+  override def save(item: Item): F[Unit] = F.delay(println(s"Saving item: $item"))
+  override def remove(name: ItemName): F[Unit] = F.delay(println(s"Removing item: $item"))
+}
+```
+
+Voilà! We got our streaming implementation of `findAll`.
+
+### Test interpreter
+
+That's all we wanted, but what about testing it? Sure, we might prefer to have a simple implementation by just using a plain `List`, so what can we possibly do?
+
+```scala
+object MemRepository extends ItemRepository[Id, List] {
+  private val mem = MutableMap.empty[String, Item]
+
+  override def findAll: List[Item] = mem.headOption.map(_._2).toList
+  override def find(name: ItemName): Id[Option[Item]] = mem.get(name.value)
+  override def save(item: Item): Id[Unit] = mem.update(item.name.value, item)
+  override def remove(name: ItemName): Id[Unit] = {
+    mem.remove(name.value)
+    ()
+  }
+}
+```
+
+That's pretty much it! We managed to abstract over the return type of `findAll` by just adding an extra parameter to our algebra.
+
+### About composition
+
+At this point the avid reader might have thought, what if I want to write a generic function that takes all the items (using `findAll`), applies some discounts and writes them back to the DB (using `save`)? Let's try and find out!
+
+```scala
+class DiscountProcessor[F[_], G[_]: Functor](repo: ItemRepository[F, G]) {
+
+  def process(discount: Double): F[Unit] = {
+    val rs: G[Item] = repo.findAll.map(item => item.copy(price = item.price * (1 - discount)))
+    // And now what? repo.save returns F[Unit]
+  }
+
+}
+```
+
+First of all, there's a fundamental problem. We are assuming that all the items are a finite number that fit into memory.
```

Comment:
I don't follow where we have made this assumption. With concrete types we certainly can find all `Item`s, modify them, and write them back (e.g. via `evalMap`) in one `Stream` without collecting all `Itme`s into memory.

Reply:
That's right. I was thinking about List and Traverse specifically when I wrote that sentence but now I see it's in the wrong position and does not make sense, will remove it.